Nvmeof/provisioner add controller publish volume

gadididi opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

This PR implements `ControllerPublishVolume` and `ControllerUnpublishVolume` methods for the NVMe-oF CSI driver, completing the Host attachment/detachment workflow. The implementation moves hosts management from volume creation to the publish/unpublish lifecycle, following proper CSI patterns.
It retrieves the `HostNqn` from `NodeID()`. 

Key changes:
- Implemented `ControllerPublishVolume` to add hosts to NVMe-oF subsystems when pods request volume attachment
- Implemented `ControllerUnpublishVolume` to remove hosts from subsystems during volume detachment  
- Moved host addition logic from `CreateVolume` to `ControllerPublishVolume`
- Added validation functions for publish/unpublish requests in `util ` package. 
- Created helper functions `publishResources` and `unpublishResources` for gateway operations


## Is there anything that requires special attention ##

**Host NQN Extraction**: The implementation assumes NodeID format follows `hostname@@NQN@@nqn.2014-08...` pattern. Node plugin must provide NodeID in this format.




## Related issues ##

Fixes: 
https://github.com/ceph/ceph-csi/issues/5370#issuecomment-3163939133


## Future concerns ##

- Implement host NQN validation with proper regex patterns

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

gadididi committed:
util: added validate methods

added basic ValidateControllerPublishVolumeRequest and ValidateControllerUnpublishVolumeRequest

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
nvmeof: remove AddHost

remove the AddHost GRPC and remove HostNQN field from NVMeoFVolumeData struct

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi commented:
Hi @Rakshith-R may you review please? thanks!

Rakshith-R commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -480,11 +664,11 @@ func (cs *Server) storeNVMeoFMetadata(
 func (cs *Server) getNVMeoFMetadata(
 	ctx context.Context,
-	req *csi.DeleteVolumeRequest,
+	secrets map[string]string,
```
Please move this change to a single commit.

gadididi replied:
fixed

gadididi committed:
nvmeof: change getNVMeofMetada argument

instead of getting the DeleteVolumeRequest, getNVMeofMetada gets the secrets itself.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
nvmeof: add Controller[Un]PublishVolume

implemented adding host (retrieve hostnqn from nodeID) GRPC call.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

Rakshith-R commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -162,6 +164,84 @@ func (cs *Server) DeleteVolume(
+	}
+
+	// populate publish context
+	publishContext := populatePublishContext(req, hostNqn)
```
```
	publishContext := range req.GetVolumeContext()
	publishContext[vcHostNQN] = hostNqn
```

Can we just directly add host in the same function ?

gadididi replied:
@Rakshith-R what do you mean same function? In ControllerPublishVolume?

Rakshith-R replied:
Yes, In the same flow.

gadididi replied:
@Rakshith-R I did not want to make this function too long, so I separated it. ( Like in CreateVolume function calls the createNvmeoFResources() )

Rakshith-R reviewed (commented):
Just a small nit

iPraveenParihar commented:
@Mergifyio rebase

iPraveenParihar commented:
@Mergifyio queue
